src/components/Storage/Card/Table/Table.tsx

```diff
@@ -0,0 +1,221 @@
+                      checked,
+                      indeterminate,
+                    } = event.target as HTMLInputElement;
+                    if (
+                      checked !== selection.allSelected ||
+                      indeterminate !== selection.allIndeterminate
+                    ) {
+                      selection.toggleAll();
+                    }
+                  }}
+                  aria-label="Select all"
+                />
+              </DataTableHeadCell>
+              <DataTableHeadCell>Name</DataTableHeadCell>
+              <DataTableHeadCell>Size</DataTableHeadCell>
+              <DataTableHeadCell>Type</DataTableHeadCell>
+              <DataTableHeadCell>Last Modified</DataTableHeadCell>
+            </DataTableRow>
+          </DataTableHead>
+          {files.length > 0 ? (
+            <DataTableBody>
+              {files.map((file) => (
+                <StorageFileTableRow
+                  selectedFile={selectedFile}
+                  bucket={bucket}
+                  key={file.fullPath}
+                  selection={selection}
+                  selectFile={selectFile}
+                  item={file}
+                  setPath={setPath}
+                />
+              ))}
+            </DataTableBody>
+          ) : null}
+        </DataTableContent>
+      </DataTable>
+      {files.length === 0 && <ZeroState />}
+    </div>
+  );
+};
+
+interface StorageTableRowProps extends StorageTableProps {
+  item: StorageItem;
+  bucket: string;
+  setPath: (path: string) => void;
+}
+
+function lastModified(item: StorageItem) {
+  if (item.type === 'folder') {
+    return <>&mdash;</>;
+  }
+
+  // Like in the console, display loading indicator in the last modified column.
+  if (item.uploading) {
+    return (
+      <div className={styles.columnUploading}>
+        <Spinner size={20} />
+        <span>Uploading...</span>
+      </div>
+    );
+  }
+
+  return new Date(item.updated).toLocaleDateString();
+}
+
+export const StorageFileTableRow: React.FC<StorageTableRowProps> = ({
+  item,
+  selection,
+  selectFile,
+  setPath,
+  selectedFile,
+}) => {
+  const fileNameId = randomId('file');
+
+  function navigate() {
+    if (item.type === 'file') {
+      if (selectedFile === item) {
+        selectFile(undefined);
+      } else {
+        selectFile(item);
+      }
+    } else {
+      selectFile(undefined);
+      setPath(item.fullPath);
+    }
+  }
+
+  return (
+    <DataTableRow
+      tabIndex={0}
+      onKeyDown={(e: KeyboardEvent<HTMLElement>) => {
+        if (e.key === 'Enter') {
+          navigate();
+        }
+      }}
+      onClick={navigate}
+      selected={selectedFile === item}
+      className={styles.row}
+    >
+      <DataTableCell hasFormControl>
+        <Checkbox
+          aria-labelledby={fileNameId}
+          checked={selection.isSelected(item.fullPath)}
+          onChange={(event: MouseEvent<HTMLInputElement>) => {
+            event.stopPropagation();
```

Comment:
If I'm not mistaken, click events will now bubble to `navigate` (e.g. checking a folder may navigate to it instead) -- consider moving this to a separate `onClick` handler. Please test manually if necessary.

Reply:
Oh, wow, good call